src/heaps/binomial_heap.jl

```diff
@@ -0,0 +1,392 @@
+            node = parent
+            parent = parent.parent
+        end
+    else increase_key!(node, new_val)
+    end
+end
+
+function top_element_node(heap::MutableBinomialHeap)
+    _heap = heap.rootList
+    length(_heap) == 0 && return nothing
+    it = iterate(_heap)
+    if it === nothing
+        throw(HeapBoundsError(eltype(heap),heap.nodecount))
+    end
+    @inbounds top_node = _heap[1]
+    while it !== nothing
+        elem, state = it
+        if elem.data < top_node.data 
+            top_node = elem 
+        end
+        it = iterate(_heap, state) 
+    end
+    return top_node
+end
+
+function update_handle!(c::Int, h::MutableBinomialHeap)
+    node::Union{Nothing, MutableBinomialHeapNode{T}} where T = nothing 
+    for i = 1:h.cumm_nodecount
+        node = nothing
+        try
+            node = find_node(h, i)
+            node.handle = node.handle + c
+        catch _
+            continue
+        end
+    end
+end
+
+# update_handle_flag determines how _union! is used
+# When explicitly called to merge two binomial heaps it is set true.
+function _union!(h1::MutableBinomialHeap{T}, h2::MutableBinomialHeap{T}, update_handle_flag::Bool) where T
+    _new = MutableBinomialHeap{T}()
+    update_handle_flag && update_handle!(h1.cumm_nodecount, h2)
+    l1 = h1.rootList
+    l2 = h2.rootList
+    h1.nodecount += h2.nodecount
+    h1.cumm_nodecount += h2.cumm_nodecount
+    iter1 = iterate(l1)
+    iter2 = iterate(l2)
+    while (iter1 !== nothing && iter2 !== nothing)
+        (elem1, state1) = iter1
+        (elem2, state2) = iter2
+        if( elem1.degree <= elem2.degree) 
+            push!(_new.rootList, elem1)
+            iter1 = iterate(l1, state1)
+        else
+            push!(_new.rootList, elem2)
+            iter2 = iterate(l2, state2)
+        end
+    end
+    while (iter1 !== nothing)
+        (elem1, state1) = iter1 
+        push!(_new.rootList, elem1) 
+        iter1 = iterate(l1, state1) 
+    end
+    while (iter2 !== nothing)
+        (elem2, state2) = iter2
+        push!(_new.rootList, elem2)
+        iter2 = iterate(l2, state2)
+    end
+    h1.rootList = _new.rootList
+    adjust!(h1)
+    return 
+end
+
+function push!(_heap::MutableBinomialHeap{T}, key) where T
+    handle = _heap.cumm_nodecount + 1
+    new_node = MutableBinomialHeapNode(convert(T, key), handle)
+    _insert_tree_in_heap!(_heap, new_node)
+    _heap.nodecount = _heap.nodecount + 1
+    _heap.cumm_nodecount = _heap.cumm_nodecount + 1
+    return new_node.handle
+end
+#################################################
+#
+#   interfaces
+#
+#################################################
+length(h::MutableBinomialHeap) = h.nodecount
+
+isempty(h::MutableBinomialHeap) = isempty(h.rootList)
+
+"""
+_union!(h1::MutableBinomialHeap, h2::MutableBinomialHeap)
+
+merges heap `h2` into heap `h1`
+"""
+_union!(h1::MutableBinomialHeap, h2::MutableBinomialHeap) = _union!(h1, h2, true)
```

Comment:
we can't have part of the public API prefixed with `_`.
That is a common convention for "Not part of public API`

We should just call this `Base.union!`